waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -1689,6 +1703,77 @@
                                   (assoc-in state [:rs-metadata :version :watch] version))))))}
       (merge options))))
 
+(defn events-by-namespace-and-workload-query
+  "Query K8s for all namespace-scoped events with matching workload names"
+  [{:keys [api-server-url] :as scheduler} options namespace workload-name]
+  (let [events-url (str api-server-url "/api/v1/namespaces/" namespace "/events?fieldSelector=involvedObject.name%3D" workload-name)
+        {:keys [items version]} (global-state-query scheduler options "get-events" events-url)
+        events (map k8s-event->simple-event items)]
+    {:events events
+     :version version}))
+
+(defn get-events-by-namespace-and-workload
+  "Load events from cache or query K8s"
+  [workload->event-cache namespace workload-name k8s-query-fn scheduler options]
+  (let [workload {:namespace namespace :workload-name workload-name}]
+    (cu/cache-get-or-load workload->event-cache workload
+                          (fn load-event-from-k8s [] (k8s-query-fn scheduler options namespace workload-name)))))
+
+(defn service-unhealthy?
+  "Return true if the given service appears unhealthy"
+  [[_ {:keys [instances task-count task-stats]}]]
+  (let [healthy-count (:healthy task-stats)]
+    (or (< task-count instances)
+        (< healthy-count task-count))))
+
+(defn add-events-to-service
+  "Add K8s Events to the given service (may acquire events from local cache or from K8s API)"
+  [service event-loader]
+  (let [namespace (:k8s/namespace service)
+        name (:k8s/app-name service)
+        events (event-loader namespace name)]
+    (assoc service :k8s/events (:events events))))
+
+(defn add-events-to-pod
+  "Add K8s Events to the given pod (may acquire events from local cache or from K8s API)"
+  [{:keys [metadata] :as pod} event-loader]
+  (let [namespace (:namespace metadata)
+        name (:name metadata)
+        events (event-loader namespace name)]
+    (assoc pod :k8s/events (:events events))))
+
+(defn start-event-fetcher!
+  "Start a thread to continuously check for problematic services and load related K8s Events for them."
+  [{:keys [watch-state workload->event-cache] :as scheduler}
+   {:keys [exit-on-error? fetch-events-chan] :as options}]
+  (doto
+   (Thread.
+    (fn fetch-k8s-events []
+      (try
+        (log/info "starting K8s event fetcher")
+        (let [event-getter (fn [namespace name] (get-events-by-namespace-and-workload workload->event-cache namespace name events-by-namespace-and-workload-query scheduler options))]
+          ;; retry forever
+          (while true
+            (log/info "K8s event fetcher checking for targets")
+            (let [{:keys [service-id->service service-id->pod-id->pod]} @watch-state]
+              (doseq [[service-id service] (filter service-unhealthy? service-id->service)]
```

Comment:
To replace what, the `filter`? What benefit would that have?

Reply:
Ah, filterm returns a map instead of a (lazy) sequence like filter does. In this case that's not needed, though, since the map (or sequence) is just iterated over.